File: Coding_1103 Trees_and_Tries/최원준/이강훈/4번.java
/*
https://leetcode.com/problems/find-duplicate-subtrees/

아이디어
DFS 로 최하위부터 상위로 String PreOrder 조합 ( Root 가 가장 앞 , Null 또한 포함 , ',' 로 각 값 구분 ) -> String 을 Key 로 숫자 카운트
 -> 기존에 있던 조합인 경우 정답에 포함, 또 추가가 안되도록 value 를 1 증가 시킴

자료구조
HashMap

시간복잡도
O(N)
*/

import java.util.*;

public class Solution {
    List<TreeNode> answer = new ArrayList<>();

    Map<String, Integer> map = new HashMap<>();

    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        dfs(root);

        return answer;
    }

    public String dfs(TreeNode node) {
        if (node == null) return "null";

        String str = String.valueOf(node.val) + "," + dfs(node.left) + "," + dfs(node.right);

        if (!map.containsKey(str)) map.put(str, 1);
        else if (map.get(str) == 1) {
            answer.add(node);
            map.put(str, 2);
        }

        return str;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: Coding_231103 Trees_and_Tries/최원준/최원준/4번.py
# https://leetcode.com/problems/find-duplicate-subtrees/description/

'''
1. 아이디어 :
    문자열을 만들어서 dfs를 돌리면 된다.
2. 시간복잡도 :
    O(n)
3. 자료구조 :
    해시맵, 배열
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:

        subtrees = defaultdict(list)

        def dfs(node):
            if not node:
                return "null"
            s = ",".join([str(node.val), dfs(node.left), dfs(node.right)])
            if len(subtrees[s]) == 1:
                res.append(node)
            subtrees[s].append(node)
            return s

        res=[]
        dfs(root)
        return res
